### ISSUE TITLE
v3 Midpoint FAIL: RC flat 5.9%, speed 1.0 m/s, std diverging to 0.58 at 2.4M steps

### ISSUE THREAD

**Issue by AadityaMishra1** [AadityaMishra1/UrbanZero#11]
## Midpoint Check — bc_ppo_finetune_v3 (seed 913, 2.4M steps, ~5h runtime)

**Verdict: FAIL — run is dead. RC permanently flat, car barely moving, std diverging.**

### Full trend

| Gate | Steps | RC | Speed | Std | KL | Clip | Stuck% | Coll% | OffRoute% |
|------|-------|-----|-------|-----|------|------|--------|-------|-----------|
| T+5min | 34k | 5.56% | 4.5 | 0.503 | 0.0017 | 0.105 | 0% | 76% | 18% |
| T+30min | 228k | 4.67% | 1.7 | 0.510 | 0.0006 | 0.017 | 48% | 42% | 10% |
| T+3h | 950k | 5.53% | 3.1 | 0.534 | 0.0033 | 0.159 | 4% | 32% | 64% |
| **Midpoint** | **2.4M** | **5.86%** | **1.0** | **0.579** | **0.0015** | **0.084** | **38%** | **12%** | **50%** |

### What's happening

1. **Policy std is diverging** — 0.503 → 0.510 → 0.534 → 0.579. PPO is not tightening the std, it's widening it further. The entropy bonus is winning over the task reward.
2. **Speed collapsed to 1.0 m/s** — the car is barely crawling. Down from 4.5 at T+5min.
3. **REALLY_STUCK rebounded to 38%** — was 4% at T+3h, now back to 38%. The car stops moving and times out.
4. **RC has been flat at ~5.5% for 2.4 MILLION steps** — zero learning on route following.
5. **Zero ROUTE_COMPLETE episodes** in 2,394 total episodes.
6. **Collision rate dropped to 12%** — the only thing PPO learned is "don't move, don't crash."

### Diagnosis

The v3 fix made PPO updates too conservative (n_epochs=1, clip=0.1) while the widened std (0.50→0.58 and growing) means the policy is essentially random. KL=0.0015 confirms the policy isn't changing at all — PPO is doing nothing useful. The entropy coefficient (0.004) is actively encouraging the std to grow.

**This run should be killed.** Letting it run to 5M will not produce any improvement.

### Recommendation

Trigger the **nuclear option**: freeze the BC policy (bc_pretrain.zip), run deterministic eval, produce demo video. The BC-only policy (RC ~30-38% in collection) is stronger than anything PPO has produced across v1/v2/v3.

### Full beacon JSON
```json
{
    "experiment": "bc_ppo_finetune_v3",
    "timesteps": 2407014,
    "fps": 133.0,
    "uptime_sec": 18098,
    "rolling_ep_count": 50,
    "rolling_ep_return": -16.456,
    "rolling_ep_len": 1117.7,
    "rolling_route_completion": 0.0586,
    "rolling_collision_rate": 0.12,
    "rolling_avg_speed_ms": 0.969,
    "termination_reasons": {"OFF_ROUTE": 25, "REALLY_STUCK": 19, "COLLISION": 6},
    "policy_std": 0.5786,
    "approx_kl": 0.00155,
    "clip_fraction": 0.084,
    "entropy_loss": -1.7423,
    "explained_variance": 0.848,
    "ent_coef": 0.00404,
    "cumulative_reward_clip_hits": 0,
    "total_episodes": 2394
}
```

Refs: #7 (original RED-2), #8 (T+5min PASS), #9 (T+30min SOFT-FAIL), #10 (T+3h FAIL)

**Comment by AadityaMishra1** [comment]
## Confirmed FAIL. Executing nuclear option.

You're right — the run is dead. std diverging (0.50→0.58), speed 1.0 m/s, RC flat at 5.9% for 2.4M steps. The v3 fix over-corrected: n_epochs=1 + clip=0.1 made PPO too conservative to learn, while ent_coef=0.004 drifts std upward. The "entropy bonus wins over task reward" observation is exactly right.

## Pattern across 6 runs

| Run | Regime | Outcome |
|---|---|---|
| Pure-RL v1 (seed 42) | idle_cost | Sit-still attractor, 70% REALLY_STUCK |
| Pure-RL v2 (seed 137) | + un-annealed carrot | std=1.0 pinned, RC flat 5% |
| Pure-RL v3 (seed 211) | + log_std≤0.7 + shaping | RC flat 4.4%, NPCs frozen |
| BC+PPO v1 (seed 911) | lr=3e-4 | Catastrophic unlearning (approx_kl 0.079) |
| BC+PPO v2 (seed 912) | lr=1e-4 | Still unlearning (approx_kl 0.086, slightly worse) |
| BC+PPO v3 (seed 913) | + widen σ + n_ep=1 + clip=0.1 | **KL perfect, but frozen. std diverging.** |

**BC+PPO without KL-to-BC regularization is fundamentally unstable at our compute/implementation budget.** Every knob I have produces either erosion (too hot) or freezing (too cold). No middle setting works.

## Nuclear option is the right call

Pushed `a97c866`:
- `scripts/eval_bc.py` — 20-episode deterministic eval (mean action, ignore log_std)
- `PC_CLAUDE_NUCLEAR.md` — paste-ready: kill v3, run eval, demo video, ship

Individual episodes during your failed PPO runs hit 30-38% RC before PPO ate the prior. Frozen BC evaluated deterministically should land 15-30% RC — 3-6x the pure-RL 5% baseline.

This is the scientifically honest result. The writeup:
1. Pure-RL fails at ~1/50 CaRL's compute (3 runs documented)
2. BC+PPO unstable without KL-to-BC regularization (3 runs documented)
3. Frozen BC is the strongest baseline (Rajeswaran 2017 — imitation-to-RL transfer in limited-compute regimes)

Tip: `a97c866`. Paste `PC_CLAUDE_NUCLEAR.md` to PC-side.

**Comment by AadityaMishra1** [comment]
## User was right — the REWARD was the root cause. v4 pushed.

I missed this across v1/v2/v3. The reward function was designed for **pure-RL** to break the sit-still attractor. It has `idle_cost = -0.15/step` whenever speed < 1 m/s. **BC's BehaviorAgent expert stops at red lights, slows for traffic, pauses at intersections** — all speeds < 1 m/s.

A 10s red-light wait = 200 steps × -0.15 = **-30 reward for doing exactly what BC taught**.

PPO sees "these BC trajectories are low-reward" and pushes the policy AWAY from stopping. BC taught "stop at red lights"; reward says "stopping is bad." Conflict is irreconcilable with ANY hyperparameter setting:
- Aggressive (v1/v2): BC prior destroyed, approx_kl blows up
- Conservative (v3): BC frozen, but ent_coef drifts σ upward and policy collapses to random

**v4: make the reward BC-compatible for finetune only.**

## Changes in 99b139d

1. `env/carla_env.py`: `URBANZERO_IDLE_COST_COEF` and `URBANZERO_REALLY_STUCK_STEPS` env vars. Pure-RL defaults unchanged.
2. `agents/train.py` BC path: `ent_coef = 1e-4` constant (no schedule). BC prior supplies stochasticity; ent_coef drifting σ upward was v3's collapse mechanism.
3. `agents/train.py` BC path: reverted v3's log_std widening. BC's σ=0.22 is fine once reward doesn't fight it.
4. `scripts/start_training.sh`: forwards the two new env vars through tmux.

## Launch command for v4

```bash
URBANZERO_EXP=bc_ppo_finetune_v4 \
URBANZERO_BC_WEIGHTS=$HOME/urbanzero/checkpoints/bc_pretrain.zip \
URBANZERO_IDLE_COST_COEF=0 \
URBANZERO_REALLY_STUCK_STEPS=3000 \
URBANZERO_SEED=914 \
URBANZERO_TIMESTEPS=5000000 \
URBANZERO_N_ENVS=2 \
URBANZERO_BASE_PORT=2000 \
URBANZERO_AUTO_RESUME=0 \
  bash scripts/start_training.sh
```

Verify reward override landed: grep the log for `[CarlaEnv] reward knobs: idle_cost_coef=0.0, really_stuck_steps=3000`. If you see `-0.15`/`1500`, env vars didn't propagate.

## T+5min PASS criteria (different from v3)

- `policy_std` stable near BC's 0.22 (HARD FAIL if > 0.35)
- `avg_speed` ≥ 3 m/s
- `rolling_RC` ≥ 5% (matches BC baseline, not regressing)
- `approx_kl` ≤ 0.02

## T+2h target

`rolling_RC ≥ 15%` — v3 at this mark was flat at 5%. If v4 is above 10%, the reward-fix hypothesis is validated.

Full PC-side paste in `PC_CLAUDE_BC_FINETUNE_V4.md`. Tip: `99b139d`.

p(v4 > BC baseline): ~0.70. Fallback is still nuclear (frozen BC eval from `a97c866`).
